loopy/transform/data.py

```diff
@@ -650,14 +653,45 @@ def rename_argument(kernel, old_name, new_name, existing_ok=False):
 
     kernel = smap.map_kernel(kernel)
 
+    # }}}
+
+    # {{{ args, temporary_variables
+
+    from loopy.kernel.array import ArrayBase
+    subst_mapper = SubstitutionMapper(make_subst_func(subst_dict))
+
     new_args = []
     for arg in kernel.args:
         if arg.name == old_name:
             arg = arg.copy(name=new_name)
+        if isinstance(arg, ArrayBase) and arg.shape:
+            arg = arg.copy(shape=subst_mapper(arg.shape))
```

Comment:
`RuleAwareIdentityMapper` expects an instance of ExpansionState to go alone with any expression which is not ideal for mapping the expressions in shapes/strides.

Reply:
I don't think that's a big issue. An instance of `ExpansionState` can be manufactured easily, as is done e.g. [here](https://github.com/inducer/loopy/blob/3742be4f68021e058b8c0c6cd4ea1fe0bd53f6b1/loopy/symbolic.py#L986-L990).

Reply:
And in this case that would  be taking in dummy 'instruction' argument which may or may not have side effects based on how the sub-classes use it. RuleAwareSubstitutionMapper wouldn't show any side effects, but there's no guarantee that all the subclasses of RuleAwareIdentityMapper would be side-effect free.

Reply:
One option could be: `if expn_state is None` for any mapper method of `RuleAwareIdentityMapper` it should just call the corresponding method of IdentityMapper.

Reply:
> the corresponding method of IdentityMapper.

How would that be accessible? Also why is it sensible, given that the behavior we want lives in a subclass (of `RuleAwareIdentityMapper`)?

Reply:
Agreed, that my comment in https://github.com/inducer/loopy/pull/155#discussion_r504989942 was ill formed.

But I'm still of the opinion that the way RuleAwareIdentityMapper is structured bars it to be an IdentityMapper for expressions not coming from a loopy kernel's instruction. [cf [RuleAwareIdentityMapper's entrypoint method](https://github.com/inducer/loopy/blob/3742be4f68021e058b8c0c6cd4ea1fe0bd53f6b1/loopy/symbolic.py#L981)]

Reply:
I see what you're saying, but it would also seem weird to have to implement all sorts of mappers twice. How do you feel about just documenting that this use is OK (i.e. that `ExpansionState` may have `instruction is None`)?

Reply:
Agreed and implemented it. Didn't turn out to be intrusive as well!

Reply:
I'm reneging on my previous statement that said the changes won't be intrusive. Precisely because `RuleAwareSubstituionMapper` in its current state is ill-defined.
1. If an expression is not within an instruction then it mimics an IdentityMapper. [cf [this line](https://github.com/inducer/loopy/blob/3dfcfed28087cfaed9211abd86972612e5d3fc44/loopy/symbolic.py#L1015-L1028)]
2. However in the past, RuleAwareSubstitutionMapper has been used as a substitution mapper for expressions not coming from instructions [see [this line](https://github.com/inducer/loopy/blob/3dfcfed28087cfaed9211abd86972612e5d3fc44/loopy/transform/precompute.py#L834)] by abusively setting the value of `within=All()`.

Reply:
>     1. If an expression is not within an instruction then it mimics an IdentityMapper. [cf [this line](https://github.com/inducer/loopy/blob/3dfcfed28087cfaed9211abd86972612e5d3fc44/loopy/symbolic.py#L1015-L1028)]

I disagree with your read here. `All` (what you get from `parse_match(None)`) will return `within() == True` for potential matches from instructions and non-instructions.

I agree that it's hokey and not well-defined (which we should fix!), but at least the two uses that you highlight are IMO actually consistent.

Reply:
Agreed. e6d7d6b